chain-signatures/contract/src/update.rs

```diff
@@ -69,33 +93,60 @@ impl ProposedUpdates {
     ///
     /// Returns Some(votes) if the given [`UpdateId`] exists, otherwise None.
     pub fn vote(&mut self, id: &UpdateId, voter: AccountId) -> Option<&HashSet<AccountId>> {
-        let votes = self.votes.get_mut(id)?;
-        votes.insert(voter);
-        Some(votes)
+        let entry = self.entries.get_mut(id)?;
+        entry.votes.insert(voter);
+        Some(&entry.votes)
     }
 
-    pub fn remove(&mut self, id: &UpdateId) -> Option<Vec<Update>> {
-        self.votes.remove(id);
-        self.updates.remove(id)
+    fn remove(&mut self, id: &UpdateId) -> Option<UpdateEntry> {
+        self.entries.remove(id)
     }
 
-    pub fn do_update(&mut self, id: &UpdateId, config_callback: &str, gas: Gas) -> Option<Promise> {
-        let updates = self.remove(id)?;
+    pub fn do_update(&mut self, id: &UpdateId, gas: Gas) -> Option<Promise> {
+        let entry = self.remove(id)?;
 
         let mut promise = Promise::new(env::current_account_id());
-        for update in updates {
+        for update in entry.updates {
             match update {
                 Update::Config(config) => {
                     promise = promise.function_call(
-                        config_callback.into(),
+                        "update_config".into(),
                         serde_json::to_vec(&(&config,)).unwrap(),
                         NearToken::from_near(0),
                         gas,
                     );
                 }
-                Update::Contract(code) => promise = promise.deploy_contract(code),
+                Update::Contract(code) => {
+                    // deploy contract then do a `migrate` call to migrate state.
+                    promise = promise.deploy_contract(code).function_call(
```

Comment:
Does this work as a batched transaction? Will deployment fail if the function call fails?

Reply:
yeah, this ends up being a batch call where calling into `migrate` will roll back the whole entire promise with the deployed contract